gravitee-am-service/src/main/java/io/gravitee/am/service/impl/CertificateServiceImpl.java

```diff
@@ -180,71 +193,57 @@ public Flowable<Certificate> findAll() {
-                            }
-                            emitter.onSuccess(certificate);
-                        });
+                .flatMap(certificateSchema -> {
+                    var certificate = new Certificate();
+                    certificate.setId(RandomString.generate());
+                    certificate.setDomain(domain);
+                    certificate.setName(newCertificate.getName());
+                    certificate.setType(newCertificate.getType());
+                    certificate.setSystem(isSystem);
+                    var certificateConfiguration = objectMapper.readTree(newCertificate.getConfiguration());
+                    var fileKey = certificateSchema.getProperties()
+                            .entrySet()
+                            .stream()
+                            .filter(map -> map.getValue().getWidget() != null && "file".equals(map.getValue().getWidget()))
+                            .map(Map.Entry::getKey)
+                            .findFirst().orElse(null);
+                    if (fileKey == null || !certificateConfiguration.has(fileKey)) {
+                        return Single.error(() -> new CertificateException("Certification file is not found"));
                     }
-                });
-
-        return certificateSingle
-                .flatMap(certificate -> certificateRepository.create(certificate))
+                    try {
+                        var file = objectMapper.readTree(certificateConfiguration.get(fileKey).asText());
+                        var fileContent = Base64.getDecoder().decode(file.get(CONTENT).asText());
+                        certificate.setMetadata(Maps.newHashMap(Map.of(CertificateMetadata.FILE, fileContent)));
+                        // update configuration to set the file name
+                        ((ObjectNode) certificateConfiguration).put(fileKey, file.get(NAME).asText());
+                        newCertificate.setConfiguration(objectMapper.writeValueAsString(certificateConfiguration));
+                        certificate.setConfiguration(newCertificate.getConfiguration());
+                        certificate.setCreatedAt(new Date());
+                        certificate.setUpdatedAt(certificate.getCreatedAt());
+                        return Single.just(certificate);
+                    } catch (Exception ex) {
+                        LOGGER.error("An error occurs while trying to create certificate configuration", ex);
+                        return Single.error(() -> new TechnicalManagementException("An error occurs while trying to create a certificate", ex));
+                    }
+                })
+                .flatMap(certificate -> {
+                    try {
+                        isValid(certificate);
+                        return certificateRepository.create(certificate);
+                    } catch (Exception e) {
+                        return Single.error(e);
+                    }
+                })
```

Comment:
**suggestion**: instead of 'isValid' which return a void, maybe we should name it 'check' or 'validate' and make it returns a Certificate. In that way, we will be able to write it like this;
```suggestion
                .map(this::validate)
                .flatMap(certificateRepository::create)
```

Reply:
it will be wrong error handling: 
java.lang.NullPointerException: The single returned by the mapper is null